pkg/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -677,6 +713,102 @@ func (q *DBJobQueue) RefreshHeartbeat(token uuid.UUID) {
 	}
 }
 
+func (q *DBJobQueue) InsertWorker(arch string) (uuid.UUID, error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return uuid.Nil, err
+	}
+	defer conn.Release()
+
+	id := uuid.New()
+	_, err = conn.Exec(context.Background(), sqlInsertWorker, id, arch)
+	if err != nil {
+		q.logger.Error(err, "Error inserting worker")
+		return uuid.Nil, err
+	}
+	return id, nil
+}
+
+func (q *DBJobQueue) UpdateWorkerStatus(workerID uuid.UUID) error {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	tag, err := conn.Exec(context.Background(), sqlUpdateWorkerStatus, workerID)
+	if err != nil {
+		q.logger.Error(err, "Error updating  worker status")
+	}
+	if tag.RowsAffected() != 1 {
+		q.logger.Error(nil, "No rows affected when refreshing updating status")
+		return jobqueue.ErrWorkerNotExist
+	}
+	return nil
+}
+func (q *DBJobQueue) Workers(olderThan time.Duration) ([]uuid.UUID, error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Release()
+
+	rows, err := conn.Query(context.Background(), sqlQueryWorkers, olderThan.String())
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	workerIDs := make([]uuid.UUID, 0)
+	for rows.Next() {
+		var w uuid.UUID
+		err = rows.Scan(&w)
+		if err != nil {
+			// Log the error and try to continue with the next row
+			q.logger.Error(err, "Unable to read token from heartbeats")
+			continue
+		}
+		workerIDs = append(workerIDs, w)
+	}
+	if rows.Err() != nil {
+		q.logger.Error(rows.Err(), "Error reading tokens from heartbeats")
+		return nil, rows.Err()
+	}
+
+	return workerIDs, nil
+}
+
+func (q *DBJobQueue) DeleteWorker(workerID uuid.UUID) error {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	var count int
+	err = conn.QueryRow(context.Background(), sqlQueryHeartbeatsForWorker, workerID).Scan(&count)
+	if err != nil {
+		return err
+	}
+
+	// If worker has any active jobs, refuse to remove the worker
+	if count != 0 {
+		return jobqueue.ErrActiveJobs
```

Comment:
What happens here if there is a job that is stuck? Will the worker be stuck too or do we have a way of dealing with that?

Reply:
So this is just the heartbeats. Only jobs which are running get inserted into heartbeats, meaning that if this job stops responding it will get removed from the heartbeats, and thus the worker will not have this job attached to it anymore.

Jobs which are stuck in pending aren't inserted into heartbeats (yet), so those also don't count.

Reply:
Cool thanks for the clarification :smiley: